File: AI/node.cpp
#include "node.hpp"
#include "node.hpp"
#include "darklogic.h"
#include "ai.h"
#include <ctime>
#include <random>
#include <iostream>

AI* Node::s_ai = nullptr;
double Node::s_cste = sqrt(2);

Node::Node(AI& ai):
	m_actionId(0), m_threadId(std::numeric_limits<unsigned char>::max()), m_depth(0), m_value(VAL_INIT), m_subValue(SUBVAL_INIT), m_isLoss(false)
{
	s_ai = &ai;
	std::srand(std::time(nullptr));
}

Node::Node(const Id& actionId_, const unsigned char& threadId_, const unsigned short& depth_):
	m_actionId(actionId_), m_threadId(threadId_), m_depth(depth_), m_value(VAL_INIT), m_subValue(SUBVAL_INIT), m_isLoss(false)
{
}

Node::Id Node::actionId() const
{
	return m_actionId;
}

unsigned char Node::threadId() const
{
	return m_threadId;
}

unsigned short Node::depth() const
{
	return m_depth;
}

unsigned short Node::value() const
{
	return m_value;
}

unsigned short int Node::subValue() const
{
	return m_subValue;
}

unsigned short int Node::realValue() const
{
	return (m_value < VAL_INIT)? m_value : (m_value + m_subValue);
}

bool Node::isLoss() const
{
	return m_isLoss;
}

unsigned short int Node::getValueFromAction(const unsigned short int action) const
{
	return m_sons.at(action)->value();
}

unsigned short int Node::getRealValueFromAction(const unsigned short int action) const
{
	return m_sons.at(action)->realValue();
}

void Node::pushCrtAction(const Id& actionId, const unsigned char threadIdx)
{
	auto it = m_sons.find(actionId);
	if (it != m_sons.end())
	{
		//std::cout << "[DEBUG] action " << actionId << " will be handled by thread " << static_cast<unsigned int>(threadIdx) << std::endl;
		it->second->setThreadIdx(threadIdx);
	}
	else
	{
		//std::cout << "[DEBUG] new action " << actionId << " will be handled by thread " << static_cast<unsigned int>(threadIdx) << std::endl;
		m_sons[actionId] = std::make_unique<Node>(actionId, threadIdx, m_depth + 1);
		m_sons[actionId]->eval(0);
	}
}

void Node::setThreadIdx(const unsigned char threadIdx)
{
	m_threadId = threadIdx;
	for (auto& subNode : m_sons)
	{
		subNode.second->setThreadIdx(threadIdx);
	}
}

void Node::eval(const unsigned char threadIdx)
{
	// play crt move
	N_DarkLogic::DarkLogic::apply(threadIdx, m_actionId);

	// check if it is a node which leads to loss
	if (N_DarkLogic::DarkLogic::isAlreadyPlayed(threadIdx))
	{
		m_value = VAL_MAX;
	}
	else if (!N_DarkLogic::DarkLogic::canBeDemonstrated(threadIdx))
	{
		m_value = VAL_MAX;
		if (!N_DarkLogic::DarkLogic::isEvaluated(threadIdx))
		{
			m_isLoss = true;
		}
	}
	// check if it is a node which leads to win
	else if (N_DarkLogic::DarkLogic::isDemonstrated(threadIdx))
	{
		m_value = 0;
		// stop reflexion because AI found a demonstration
		s_ai->stopThread(threadIdx);
	}
	else
	{
		m_subValue = s_ai->eval({ N_DarkLogic::DarkLogic::getState(threadIdx) }, threadIdx);
	}

	// unplay crt move
	N_DarkLogic::DarkLogic::unapply(threadIdx);
}

unsigned short Node::exploreBasic(const std::vector<N_DarkLogic::Action::Id>& actions)
{
	unsigned short maxDepth = minDepth();
	unsigned char threadId = m_sons[actions[0]]->threadId();
	//std::chrono::high_resolution_clock::time_point start, end;
	//start = std::chrono::high_resolution_clock::now();
	while (!s_ai->mustStop(threadId))
	{
		for (const auto& action : actions)
		{
			auto node = m_sons[action].get();
			unsigned short retValue = USHRT_MAX;
			if (node->value() < USHRT_MAX)
			{
				node->exploreBasic(maxDepth);
			}

			if ((m_value > retValue + 1))
			{
				m_value = retValue + 1;
			}

			//if must stop exploration, stop it
			if (s_ai->mustStop(threadId))
			{
				break;
			}
		}
		/*if (maxDepth >= 3)
		{
			end = std::chrono::high_resolution_clock::now();
			double elapsed_seconds = std::chrono::duration<double>(end - start).count();
			std::cout << "[DEBUG] thread id: " + numberToString(static_cast<unsigned int>(threadId)) + " has finished depth '" + numberToString(maxDepth) <<
				"' in " + numberToString(elapsed_seconds) + " seconds" << std::endl;

		}*/
		maxDepth++;
	}
	return m_value;
}

unsigned short Node::exploreBasic(const unsigned short maxDepth)
{
	//play crt move
	N_DarkLogic::DarkLogic::apply(m_threadId, m_actionId);

	//no need to go deeper
	if (m_depth == maxDepth)
	{
		//check if it is a node which leads to loss
		if (N_DarkLogic::DarkLogic::isAlreadyPlayed(m_threadId) || !N_DarkLogic::DarkLogic::canBeDemonstrated(m_threadId))
		{
			m_value = USHRT_MAX;
		}
		//check if it is a node which leads to win
		else if (N_DarkLogic::DarkLogic::isDemonstrated(m_threadId))
		{
			m_value = 0;
			
			//stop reflexion because AI found a demonstration
			s_ai->stopThread(m_threadId);
		}
	}
	else if(!s_ai->mustStop(m_threadId))
	{
		//get actions
		auto actions = N_DarkLogic::DarkLogic::getActions(m_threadId);

		//add all subnodes if they have not been created yet
		if (m_depth == maxDepth - 1)
		{
			for (const auto& action : actions)
			{
				m_sons[action] = std::make_unique<Node>(action, m_threadId, m_depth + 1);
			}			
		}

		//explore subNodes
		bool hasOnlyLosses = true;
		for (const auto& action : actions)
		{
			//explore node associated with action
			auto node = m_sons[action].get();
			unsigned short retValue = USHRT_MAX;
			if (node->value() < USHRT_MAX)
			{
				retValue = node->exploreBasic(maxDepth);
			}

			//update m_value
			if (retValue == USHRT_MAX)
			{
				if (hasOnlyLosses)
				{
					m_value = USHRT_MAX;
				}
			}
			else
			{
				hasOnlyLosses = false;
				if (retValue==VAL_INIT && m_value == USHRT_MAX)
				{
					m_value = VAL_INIT;
				}
				else if ((m_value > retValue + 1))
				{
					m_value = retValue + 1;
				}
			}

			//if must stop exploration, stop it
			if (s_ai->mustStop(m_threadId))
			{
				break;
			}
		}
	}

	//unplay crt move
	N_DarkLogic::DarkLogic::unapply(m_threadId);

	return m_value;
}

void Node::exploreEval(DbAction& dbAction, unsigned char threadIdx)
{
	while (!s_ai->mustStop(threadIdx))
	{
		const auto& action = dbAction.getBestAction();
		auto node = m_sons[action].get();
		unsigned short int value = node->exploreEval();
		dbAction.updateValue(action, value);

		// if must stop exploration, stop it
		if (s_ai->mustStop(threadIdx))
		{
			break;
		}
	}
}

unsigned short Node::exploreEval()
{
	// play crt move
	N_DarkLogic::DarkLogic::apply(m_threadId, m_actionId);
	// get actions
	auto actions = N_DarkLogic::DarkLogic::getActions(m_threadId);
	//no need to go deeper
	if (m_sons.size() != actions.size())
	{
		unsigned short minSubValue = SUBVAL_INIT;
		bool updateValue = true;
		for (const auto action : actions)
		{
			Node* node = nullptr;
			if (!m_sons.contains(action))
			{
				m_sons[action] = std::make_unique<Node>(action, m_threadId, m_depth + 1);
				node = m_sons[action].get();
				node->eval(m_threadId);
				if (node->value() < VAL_MAX)
				{
					auto idx = m_dbActions.push(action);
					m_dbActions.updateValue(idx, node->realValue());
				}
			}
			else
			{
				node = m_sons[action].get();
			}
			unsigned short subValue = node->subValue();
			if (subValue < minSubValue)
			{
				minSubValue = subValue;
			}

			// if must stop exploration, stop it
			if (s_ai->mustStop(m_threadId))
			{
				if (node->value() < VAL_INIT)
				{
					m_value = node->value() + 1;
				}
				updateValue = false;
				break;
			}
		}
		if (updateValue)
		{
			if (m_dbActions.size() == 0)
			{
				//all actions lead to loss
				m_value = VAL_MAX;
			}
			else
			{
				m_subValue = minSubValue + 1;
			}
		}
	}
	else
	{
		auto action = m_dbActions.getBestAction();
		Node* node = m_sons[action].get();
		unsigned short value = node->exploreEval();
		if (node->value() < VAL_MAX)
		{
			m_dbActions.updateValue(action, value);
			action = m_dbActions.getBestAction();
			node = m_sons[action].get();
			m_subValue = node->subValue() + 1;
			if (node->value() < VAL_INIT)
			{
				m_value = node->value() + 1;
			}
		}
		else
		{
			m_dbActions.removeIdx(action);
		}
	}

	// unplay crt action
	N_DarkLogic::DarkLogic::unapply(m_threadId);

	return realValue();
}

Node* Node::getBestNode()
{
	std::vector<size_t> minNodes;
	unsigned short minVal = USHRT_MAX;
	for (auto& son : m_sons)
	{
		if (son.second->realValue() < minVal)
		{
			minVal = son.second->realValue();
			minNodes.clear();
			minNodes.push_back(son.first);
		}
		else if (son.second->realValue() == minVal)
		{
			minNodes.push_back(son.first);
		}
	}
	auto valWinner= minNodes[rand() % minNodes.size()];
	auto winner = m_sons[valWinner].release();
	m_sons.clear();	
	return winner;
}

Node* Node::getDemoMode()
{
	Node* winner = nullptr;
	std::vector<size_t> minNodes;
	unsigned short minVal = VAL_INIT-1;
	for (auto& son : m_sons)
	{
		if (son.second->value() < minVal)
		{
			minVal = son.second->value();
			minNodes.clear();
			minNodes.push_back(son.first);
		}
		else if (son.second->value() == minVal)
		{
			minNodes.push_back(son.first);
		}
	}
	if (minNodes.size())
	{
		auto valWinner = minNodes[rand() % minNodes.size()];
		winner = m_sons[valWinner].release();
		m_sons.clear();
	}	
	return winner;
}

size_t Node::nbNode() const
{
	size_t ret = 1;
	for (const auto& son : m_sons)
	{
		ret += son.second->nbNode();
	}
	return ret;
}

void Node::decrDepth()
{
	m_depth--;
	for (auto& son : m_sons)
	{
		if (son.second)
		{
			son.second->decrDepth();
		}
	}
}

unsigned short Node::minDepth() const
{
	if (m_sons.size())
	{
		unsigned short ret = USHRT_MAX;
		for (auto& son : m_sons)
		{
			if (son.second->value() < USHRT_MAX)
			{
				unsigned short crtDepth = son.second->minDepth();
				if (crtDepth + 1 < ret)
				{
					ret = crtDepth + 1;
				}
			}
			
		}
		return ret;
	}
	else
	{
		return 0;
	}
}

File: PyDarkLogic/AI/node.py
import time
import random as rand
import MainDarkLogic.darklogic as DarkLogic
from Database.state import State
from AI.dbnode import DbNode


class Node:
    """description of class"""
    # static ai
    _ai = None
    VAL_MAX = 101
    VAL_INIT = VAL_MAX - 1
    INIT_SUBVALUE = 10 ** 6

    def __init__(self, **kwargs):
        # First Root node
        if "ai" in kwargs:
            Node._ai = kwargs["ai"]
            self._actionId = 0
            self._threadId = -1
            self._depth = 0
        else:
            self._actionId = kwargs["actionId"]
            self._threadId = kwargs["threadId"]
            self._depth = kwargs["depth"]
        self._value = Node.VAL_INIT
        self._subValue = Node.INIT_SUBVALUE
        self._isEvaluated = False
        self._aiValue = None
        self._sons = {}
        self._dbNode = DbNode()
        self._isLoss = False

    def actionId(self):
        return self._actionId

    def threadId(self):
        return self._threadId

    def depth(self):
        return self._depth

    def value(self):
        return self._value

    def setValue(self, value):
        self._value = value

    def aiValue(self):
        return self._aiValue

    def setAIValue(self, aivalue):
        self._aiValue = aivalue

    def subValue(self):
        return self._subValue

    def realValue(self):
        if self._value < Node.VAL_INIT:
            return self._value
        else:
            return self._value + self._subValue

    def isEvaluated(self):
        return self._isEvaluated

    def isAIValuated(self):
        return not (self._aiValue is None)

    def isSubValuated(self):
        return self._subValue != Node.INIT_SUBVALUE

    def pushCrtAction(self, actionIds, threadIdx):
        actionId = actionIds[0]
        if len(actionIds) == 1:
            if actionId in self._sons and self._sons[actionId]:
                self._sons[actionId].setThreadIdx(threadIdx)
            else:
                self._sons[actionId] = Node(actionId=actionId, threadId=threadIdx, depth=self._depth + 1)
                # eval node
                self._sons[actionId].eval(0)
        else:
            self._sons[actionId].pushCrtAction(actionIds[1:], threadIdx)

    def eval(self, threadIdx):
        # play crt move
        DarkLogic.apply(threadIdx, self._actionId)

        # check if it is a node which leads to loss
        if DarkLogic.isAlreadyPlayed(threadIdx):
            self._value = Node.VAL_MAX
        elif not DarkLogic.canBeDemonstrated(threadIdx):
            self._value = Node.VAL_MAX
            if not DarkLogic.isEvaluated(threadIdx):
                self._isLoss = True
        # check if it is a node which leads to win
        elif DarkLogic.isDemonstrated(threadIdx):
            self._value = 0
            # stop reflexion because AI found a demonstration
            Node._ai.stopThread(threadIdx)
        else:
            self._subValue = Node._ai.eval([DarkLogic.getState(threadIdx)], threadIdx)
        self._isEvaluated = True

        # unplay crt move
        DarkLogic.unapply(threadIdx)

    def getValueOfActions(self, actions):
        action = actions[0]
        if len(actions) == 1:
            return self._sons[action].value()
        else:
            return self._sons[action].getValueOfActions(actions[1:])

    def getRealValueOfActions(self, actions):
        action = actions[0]
        if len(actions) == 1:
            return self._sons[action].realValue()
        else:
            return self._sons[action].getRealValueOfActions(actions[1:])

    def getSubValueFromSon(self, action):
        return self._sons[action].subValue()

    def setThreadIdx(self, threadIdx):
        self._threadId = threadIdx
        for subNode in self._sons.values():
            if subNode:
                subNode.setThreadIdx(threadIdx)

    def isRoot(self):
        return self._threadId == 0

    def getSubNode(self, actions):
        if len(actions) == 1:
            return self._sons[actions[0]]
        else:
            return self._sons[actions[0]].getSubNode(actions[1:])

    def exploreStatic(self, actions, threadId):
        maxDepth = self.minDepth()
        # threadId = self._sons[actions[0]].threadId()
        start = time.perf_counter()
        while not Node._ai.mustStop(threadId):
            for action in actions:
                # go to node
                node = self._sons[action]
                retValue = Node.VAL_MAX
                if node.value() < Node.VAL_MAX:
                    node.exploreDepthStatic(maxDepth)
                if self._value > retValue + 1:
                    self._value = retValue + 1

                # if must stop exploration, stop it
                if Node._ai.mustStop(threadId):
                    break

            """"if maxDepth >= 3:
                end = time.perf_counter()
                elapsed_seconds = end - start
                print("[DEBUG] thread id: " + str(threadId) + " has finished depth '" + str(maxDepth)
                      + "' in " + str(elapsed_seconds) + " seconds")"""
            maxDepth += 1

        return self._value

    def exploreDepthStatic(self, maxDepth):
        # play crt move
        DarkLogic.apply(self._threadId, self._actionId)
        """print("crt action id: " + str(self._actionId) + ", at depth="
              + str(self._depth) + ", crt theorem is: " + DarkLogic.toStrTheorem(self._threadId))"""

        # no need to go deeper
        if self._depth == maxDepth:
            # check if it is a node which leads to loss
            if DarkLogic.isAlreadyPlayed(self._threadId) or not DarkLogic.canBeDemonstrated(self._threadId):
                self._value = Node.VAL_MAX
            # check if it is a node which leads to win
            elif DarkLogic.isDemonstrated(self._threadId):
                self._value = 0
                # stop reflexion because AI found a demonstration
                Node._ai.stopThread(self._threadId)
            """else:
                self._subValue = self._ai.eval([DarkLogic.getState(self._threadId)], self._threadId)"""
        elif not Node._ai.mustStop(self._threadId):
            # get actions
            actions = DarkLogic.getActions(self._threadId)

            # add all subnodes if they have not been created yet
            if self._depth == maxDepth - 1:
                for action in actions:
                    self._sons[action] = Node(actionId=action, threadId=self._threadId, depth=self._depth + 1)

            # explore subNodes
            hasOnlyLosses = True
            self._subValue = Node.INIT_SUBVALUE
            for action in actions:
                # explore node associated with action
                node = self._sons[action]
                retValue = Node.VAL_MAX
                if node.value() < Node.VAL_MAX:
                    retValue = node.exploreDepthStatic(maxDepth)

                # update m_value
                if retValue == Node.VAL_MAX:
                    if hasOnlyLosses:
                        self._value = Node.VAL_MAX
                else:
                    hasOnlyLosses = False
                    if retValue == Node.VAL_INIT and self._value == Node.VAL_MAX:
                        self._value = Node.VAL_INIT
                    elif self._value > retValue + 1:
                        self._value = retValue + 1

                    # update subValue
                    if self._subValue > node.subValue():
                        self._subValue = node.subValue()

                # if must stop exploration, stop it
                if Node._ai.mustStop(self._threadId):
                    break

        # unplay crt move
        DarkLogic.unapply(self._threadId)

        return self._value

    def exploreDeep(self, actions):
        actionIdx = 0
        maxDepth = 0
        while True:
            # init Nodes
            states = []
            nodeList = []
            crtActions = actions[actionIdx:]

            if len(crtActions) > 0:
                for action in crtActions:
                    actionIdx += 1
                    node = self._sons[action]
                    if not node.isEvaluated():
                        node.preExplore(nodeList, states)
                        if node.value() < Node.VAL_INIT:
                            self._value = node.value() + 1
                    if Node._ai.mustStop(self._threadId):
                        return
                    if len(states) == Node._ai.MaxNbNode:
                        break

            if Node._ai.mustStop(self._threadId):
                return
            if len(states) != Node._ai.MaxNbNode:
                # We need to go deeper!
                nodes = self.getDeepExploreNodes(actions)
                rand.shuffle(nodes)
                for node in nodes:
                    node.exploreDepthDeep(nodeList, states)
                    if node.value() < Node.VAL_INIT:
                        self._value = node.value() + 1
                    if Node._ai.mustStop(self._threadId):
                        return
                    if len(states) == Node._ai.MaxNbNode:
                        break

            # evaluate nodes
            Node._ai.evaluate(nodeList, states)

    def exploreDepthDeep(self, nodeList, states):
        DarkLogic.apply(self._threadId, self._actionId)
        actions = DarkLogic.getActions(self._threadId)
        for action in actions:
            if not self._sons[action]:
                self._sons[action] = Node(actionId=action, threadId=self._threadId, depth=self._depth + 1)
            node = self._sons[action]
            if not node.isEvaluated():
                node.preExplore(nodeList, states)
                if node.value() < Node.VAL_INIT:
                    self._value = node.value() + 1
            if Node._ai.mustStop(self._threadId):
                break
            if len(states) == Node._ai.MaxNbNode:
                break

        if not Node._ai.mustStop(self._threadId) and len(states) != Node._ai.MaxNbNode:
            # We need to go deeper!
            nodes = self.getDeepExploreNodes(self._sons.keys())
            rand.shuffle(nodes)
            for node in nodes:
                node.exploreDepthDeep(nodeList, states)
                if node.value() < Node.VAL_INIT:
                    self._value = node.value() + 1
                if Node._ai.mustStop(self._threadId):
                    break
                if len(states) == Node._ai.MaxNbNode:
                    break

        DarkLogic.unapply(self._threadId)

    def exploreEval(self, dbNode, threadId):
        while not Node._ai.mustStop(threadId):
            action = dbNode.getBestAction()

            node = self._sons[action]
            value = node.exploreDepthEval()
            dbNode.updateValue(action, value)

            # if must stop exploration, stop it
            if Node._ai.mustStop(threadId):
                break

    def exploreDepthEval(self):
        # play crt move
        DarkLogic.apply(self._threadId, self._actionId)
        # get actions
        actions = DarkLogic.getActions(self._threadId)
        # no need to go deeper
        if len(self._sons) != len(actions):
            minSubValue = Node.INIT_SUBVALUE
            updateValue = True
            for action in actions:
                # node = None
                if action not in self._sons:
                    node = Node(actionId=action, threadId=self._threadId, depth=self._depth + 1)
                    self._sons[action] = node
                    node.eval(self._threadId)
                    if node.value() < Node.VAL_MAX:
                        idx = self._dbNode.push(action)
                        self._dbNode.updateValue(idx, node.realValue())
                else:
                    node = self._sons[action]
                subValue = node.subValue()
                if subValue < minSubValue:
                    minSubValue = subValue

                # if must stop exploration, stop it
                if Node._ai.mustStop(self._threadId):
                    updateValue = False
                    break
            if updateValue:
                if self._dbNode.size() == 0:
                    # all actions lead to loss
                    self._value = Node.VAL_MAX
                else:
                    self._subValue = minSubValue + 1
        else:
            action = self._dbNode.getBestAction()
            node = self._sons[action]
            # self._dbNode.updateValue(idx, self.realValue(), True)  # delete this line
            node.exploreDepthEval()
            if node.value() < Node.VAL_MAX:
                self._dbNode.updateValue(action, self.realValue())
                action = self._dbNode.getBestAction()
                node = self._sons[action]
                self._subValue = node.subValue() + 1
            else:
                self._dbNode.removeIdx(action)

        DarkLogic.unapply(self._threadId)
        return self.realValue()

    def computeAIValue(self):
        minValue = Node.VAL_MAX
        allNode = True
        for node in self._sons.values():
            if node:
                if node.isAIValuated():
                    crtValue = node.computeAIValue()
                    if minValue > crtValue:
                        minValue = crtValue
                else:
                    allNode = False
            else:
                allNode = False
        if allNode:
            if minValue != self._aiValue - 1:
                self._aiValue = minValue + 1
        return self._aiValue

    def computeValue(self):
        minValue = self._value - 1
        for node in self._sons.values():
            crtValue = node.computeValue()
            if minValue > crtValue:
                minValue = crtValue
        if minValue + 1 != self._value:
            self._value = minValue + 1
        return self._value

    def computeSubValue(self):
        minValue = Node.INIT_SUBVALUE
        allNode = True
        for node in self._sons.values():
            if node.isSubValuated():
                crtValue = node.computeSubValue()
                if minValue > crtValue:
                    minValue = crtValue
            else:
                allNode = False
        if allNode and len(self._sons) != 0:
            self._subValue = minValue + 1
        return self._subValue

    def getDeepExploreNodes(self, actions):
        minNodes = []
        minVal = Node.VAL_MAX
        for action in actions:
            node = self._sons[action]
            if not node:
                self._sons[action] = Node(actionId=action, threadId=self._threadId, depth=self._depth + 1)
                node = self._sons[action]
            if node.value() == Node.VAL_MAX:
                continue
            if node.isAIValuated():
                # print("Node is AIEvaluated!!?")
                crtValue = node.computeAIValue()
                if crtValue < minVal:
                    minVal = crtValue
                    minNodes.clear()
                    minNodes.append(node)
                elif crtValue == minVal:
                    minNodes.append(node)
            else:
                if minVal != -1:
                    minVal = -1
                    minNodes.clear()
                minNodes.append(node)

        return minNodes

    def preExplore(self, nodeList, states):
        # play crt move
        DarkLogic.apply(self._threadId, self._actionId)

        # check if it is a node which leads to loss
        if DarkLogic.isAlreadyPlayed(self._threadId):
            self._value = Node.VAL_MAX
        elif not DarkLogic.canBeDemonstrated(self._threadId):
            self._value = Node.VAL_MAX
            if not DarkLogic.isEvaluated(self._threadId):
                self._isLoss = True
        # check if it is a node which leads to win
        elif DarkLogic.isDemonstrated(self._threadId):
            self._value = 0
            # stop reflexion because AI found a demonstration
            Node._ai.stopThread(self._threadId)
        else:
            actions = DarkLogic.getActions(self._threadId)
            for action in actions:
                self._sons[action] = None
            nodeList.append(self)
            if self._ai.canEvaluate(DarkLogic.getState(self._threadId)):
                states.append(Node._ai.getTrueState(self._threadId))
            else:
                self._aiValue = Node.VAL_INIT
        self._isEvaluated = True

        # unplay crt move
        DarkLogic.unapply(self._threadId)

    def getBestNode(self):
        minNodes = []
        minVal = Node.VAL_MAX
        minSubVal = Node.INIT_SUBVALUE
        self.computeValue()
        self.computeSubValue()
        for key in self._sons.keys():
            son = self._sons[key]
            if son.value() < minVal:
                minVal = son.value()
                minSubVal = son.subValue()
                minNodes.clear()
                minNodes.append(key)
            elif son.value() == minVal:
                if son.subValue() < minSubVal:
                    minSubVal = son.subValue()
                    minNodes = [key]
                elif son.subValue() == minSubVal:
                    minNodes.append(key)
        valWinner = minNodes[rand.randint(0, len(minNodes) - 1)]
        winner = self._sons[valWinner]
        # self._sons.clear()
        return winner

    def getDeepBestNode(self):
        minNodes = []
        minVal = Node.VAL_MAX
        minAIVal = Node.VAL_MAX
        for key in self._sons.keys():
            son = self._sons[key]
            if son.value() < minVal:
                minVal = son.value()
                if son.isAIValuated():
                    minAIVal = son.aiValue()
                minNodes.clear()
                minNodes.append(key)
            elif son.value() == minVal:
                if son.isAIValuated():
                    if son.aiValue() < minAIVal:
                        minAIVal = son.aiValue()
                        minNodes = [key]
                    elif son.aiValue() == minAIVal:
                        minNodes.append(key)

        valWinner = minNodes[rand.randint(0, len(minNodes) - 1)]
        winner = self._sons[valWinner]
        # self._sons.clear()
        return winner

    def getDemoMode(self):
        winner = None
        minNodes = []
        minVal = Node.VAL_INIT - 1
        for key in self._sons:
            son = self._sons[key]
            if son:
                if son.value() < minVal:
                    minVal = son.value()
                    minNodes.clear()
                    minNodes.append(key)
                elif son.value() == minVal:
                    minNodes.append(key)
        if len(minNodes):
            valWinner = minNodes[rand.randint(0, len(minNodes) - 1)]
            winner = self._sons[valWinner]
            # self._sons.clear()
        return winner

    def setRoot(self):
        self._threadId = 0
        self._decrDepth()

    def nbNode(self):
        ret = 1
        for son in self._sons.values():
            if son:
                ret += son.nbNode()
        return ret

    def _decrDepth(self):
        self._depth -= 1
        for son in self._sons.values():
            if son:
                son._decrDepth()

    def minDepth(self):
        if len(self._sons):
            ret = Node.VAL_MAX
            for son in self._sons.values():
                if son.value() < Node.VAL_MAX:
                    crtDepth = son.minDepth()
                    if crtDepth + 1 < ret:
                        ret = crtDepth + 1
            return ret
        else:
            return 0

    def getTrainNodes(self, x, y):
        DarkLogic.getActions()
        print(str(len(self._sons)) + " nodes to explore")
        for key in self._sons:
            node = self._sons[key]
            if node:
                if node.isEvaluated():
                    print("explore " + str(node.actionId()) + " at depth = " + str(1) + " with value = " + str(
                        node.value()))
                    if node.value() != Node.VAL_INIT and not (node.value() == Node.VAL_MAX and not node.isLoss()):
                        """print("train with node id = " + str(node.actionId()) + " at depth = " + str(
                            1) + " with value = " + str(node.value()))"""
                        Node._ai.getTrainingStates(node.value(), x, y)
                    node.getDepthTrainNodes(x, y, 1)

    def getDepthTrainNodes(self, x, y, depth):
        DarkLogic.getActions()
        DarkLogic.apply(self._actionId)
        # print("explore " + str(self.actionId()) + " at depth = " + str(depth) + " with value = " + str(self.value()))
        for key in self._sons:
            node = self._sons[key]
            if node:
                if node.isEvaluated():
                    if node.value() != Node.VAL_INIT and not (node.value() == Node.VAL_MAX and not node.isLoss()):
                        """print("train with node id = " + str(node.actionId()) + " at depth = " + str(
                            depth + 1) + " with value = " + str(node.value()))"""
                        Node._ai.getTrainingStates(node.value(), x, y)
                    node.getDepthTrainNodes(x, y, depth + 1)
        DarkLogic.unapply()

    def isLoss(self):
        return self._isLoss

    def getDbStates(self):
        # print("get all nodes from current search tree")
        ret = []
        DarkLogic.getActions()
        name = DarkLogic.theoremName()
        content = DarkLogic.toNormStrTheorem()
        if self.value() != Node.VAL_INIT and not (self.value() == Node.VAL_MAX
                                                  and not self.isLoss()):
            ret.append(State(name=name, content=content, value=self._value))
        else:
            ret.append(State(name=name, content=content))
        for key in self._sons:
            node = self._sons[key]
            if node:
                node.getDeepDbStates(ret, 1)
        return ret

    def getDeepDbStates(self, dbStates, depth):
        DarkLogic.getActions()
        # print("apply action :"+str(self._actionId)+" at depth = "+str(depth))
        DarkLogic.apply(self._actionId)
        name = DarkLogic.theoremName()
        # print("name:" + str(name))
        # print("content: "+DarkLogic.toStrTheorem())
        content = DarkLogic.toNormStrTheorem()
        # print("NormContent = " + str(content))
        if self.value() != Node.VAL_INIT and not (self.value() == Node.VAL_MAX
                                                  and not self.isLoss()):
            dbStates.append(State(name=name, content=content, value=self._value))
        else:
            dbStates.append(State(name=name, content=content))
        for key in self._sons:
            node = self._sons[key]
            if node:
                node.getDeepDbStates(dbStates, depth + 1)
        DarkLogic.unapply()
